### DISCUSSION TITLE
Script to share: MapUnit Symbology Colors to DMU RGB

### DISCUSSION THREAD

**Discussion by sarah-francis-MGS** [usgs/gems-tools-pro#125] [Scripts]
If you're tired of manually filling in RGB codes in your DMU tables, this script is for you.

This script reads layer color symbology for geology polygons and adds RGB codes to the AreaFillRGB field in the description of map units table. If the layer symbology is not in RGB, it will convert the color codes to RGB based on input color mode. It works with RGB, HSV, HSL, and CMYK. It will NOT change the color mode of the layer symbology, but it will insert the corresponding RGB codes into the DMU. 

The original code structure was put together by Evan, I just expanded it out. The zip file includes the python script and an .atbx file with the script tool configurations and metadata. To use, unzip the folder, add the .atbx file to your ArcGIS Pro project, and verify in the script properties that the path to the .py file is correct. Tool metadata has been set up, so you should be able to hover over the red asterisk/"i" icon to view parameter information.

Let me know if you have any problems with it. I'm self-taught in Python so I'm always open to feedback.

[MapUnitSymbologyColors_to_DMU_RGB.zip](https://github.com/user-attachments/files/18041314/MapUnitSymbologyColors_to_DMU_RGB.zip)

**Comment by sarah-francis-MGS** [comment]
It looks like Evan expanded this out into a [Jupyter Notebook](https://github.com/DOI-USGS/gems-tools-pro/blob/notebooks/Populate%20AreaFillRGB.ipynb) too! If you have some knowledge of Python/Jupyter notebooks, this offers a bit more flexibility in how you assign colors.

**Comment by aldunn4** [comment]
I am getting this error when trying to run the tool. Can you provide any assistance?

I'm attempting to use CMUMapUnitPolys which are in CMYK with no label split character.

`﻿Traceback (most recent call last):
﻿  File "O:\Style_and_Standards\DataModels\GeMS\MapUnitSymbologyColors_to_DMU_RGB\MapUnitSymbologyColors_to_DMU_RGB.py", line 98, in <module>
﻿    l = m.listLayers(input_layer)[0]
﻿        ~~~~~~~~~~~~~~~~~~~~~~~~~^^^
﻿IndexError: list index out of range`

**Reply by sarah-francis-MGS** [reply]
Hmm, I have not seen that error before. Lines 96-98 access an existing layer in a map in Pro. It works best if you click and drag a layer from your map into the "input layer" parameter on the geoprocessing tool, and keep that same map active when you run the tool.  If you have a different map active, it will not find the layer you are looking for. 

The line `l = m.listLayers(input_layer)[0]` looks for a layer in your active map that matches the name you've put into the "input_layer" parameter. The `[0]` accesses the first matching layer, of which there should only be one. If it cannot find any matching layers, it will tell you that your list index is out of range.

I hope this helps!!

**Reply by aldunn4** [reply]
I got it to work; I had the CMUMapUnitPolys in a QuadNameCMU group layer, and once I pulled it out on its own, it worked.

**Reply by sarah-francis-MGS** [reply]
Great, glad it worked! That makes sense that it wouldn't work in a group layer.

**Comment by sarah-francis-MGS** [comment]
A few people (including me) have gotten the following error:
![unnamed](https://github.com/user-attachments/assets/1cab238e-526b-4dff-bfbd-cc9662d308c5)
To fix this, open the Python script and change the number 1 to a number 0 in the line that threw the error. It should now read: `color_list = mapunit.symbol.symbol.symbolLayers[0].color.values`
(make sure not to change any indentation, just change the 1 to a 0)

I'm not really sure why this change fixed it, but hopefully it fixes it for you too!
